test/test_auth.py

```diff
@@ -201,3 +202,216 @@ def test_scram_both(bouncer):
+def test_auth_dbname_usage_with_target_db(
+    bouncer,
+):
+    """Regression for https://github.com/pgbouncer/pgbouncer/issues/314
+
+    Check that we handle correctly attempts to use pgbouncer DB as an auth_dbname.
+    Conditions:
+        * auth_query is set
+        * stats_users is set and does not exist in the userlist.txt
+        * client connects to pgbouncer (admin DB)
+        * global auth_dbname is not set
+        * target client DB is using as an auth_dbname
+    """
+
+    config = f"""[databases]
+        [pgbouncer]
+        auth_query = SELECT usename, passwd FROM pg_shadow where usename = $1
+        auth_user = pswcheck
+        stats_users = stats
+        listen_addr = {bouncer.host}
+        admin_users = pswcheck
+        auth_type = md5
+        auth_file = {bouncer.auth_path}
+        listen_port = {bouncer.port}
+        logfile = {bouncer.log_path}
+    """
+
+    # good password
+    bouncer.test(user="pgbouncer", password="fake")
+
+    with bouncer.log_contains(
+        'cannot use the reserved "pgbouncer" database as an auth_dbname', 1
+    ):
+        with bouncer.run_with_config(config):
+            with pytest.raises(psycopg.OperationalError, match="bouncer config error"):
+                # bad password
+                # We expect that stats user does not exist in userlist.txt
+                bouncer.test(user="stats", password="stats", dbname="pgbouncer")
+
+
+@pytest.mark.skipif("WINDOWS", reason="Windows does not have SIGHUP")
+def test_auth_dbname_usage_using_autodb(
+    bouncer,
+):
+    """Regression for https://github.com/pgbouncer/pgbouncer/issues/314
+
+    Check that the pgbouncer does not apply config which contains
+    explicitly pgbouncer database (admin DB) set in auto-database
+    definition
+    """
+
+    config = f"""
+        [databases]
+        * = host={bouncer.host} port={bouncer.port} auth_dbname=pgbouncer
+        [pgbouncer]
+        auth_query = SELECT usename, passwd FROM pg_shadow where usename = $1
+        auth_user = pswcheck
+        stats_users = stats
+        listen_addr = {bouncer.host}
+        admin_users = pswcheck
+        auth_type = md5
+        auth_file = {bouncer.auth_path}
+        listen_port = {bouncer.port}
+        logfile = {bouncer.log_path}
+    """
+
+    with bouncer.log_contains(
+        'cannot use the reserved "pgbouncer" database as an auth_dbname', 1
+    ):
+        with bouncer.run_with_config(config):
+            pass
+
+
+@pytest.mark.skipif("WINDOWS", reason="Windows does not have SIGHUP")
+def test_auth_dbname_usage_using_global_setting(
+    bouncer,
+):
+    """Regression for https://github.com/pgbouncer/pgbouncer/issues/314
+
+    Check that the pgbouncer does not apply config which contains
+    explicitly "pgbouncer" database (admin DB) set in [pgbouncer] section
+    """
+
+    config = f"""
+        [databases]
+        * = host={bouncer.host} port={bouncer.port} auth_dbname=pgbouncer
```

Comment:
if the goal of this test to see if we can detect [pgbouncer]/auth_dbname misconfiguration, then it would be nicer to avoid providing incorrect auth_dbname in the [databases] section.

Reply:
The current implementation checks the `auth_dbaname` in the `set_autodb` function and had a different parsing scenario in comparison with the common DB definition. Relying on your comment [here](https://github.com/pgbouncer/pgbouncer/pull/817#discussion_r1161149558) I will keep only one test for [databases] section because if we will use only one loop for DBs (even for autodb), we check the same scenario.

Reply:
I think @goksgie meant you should probably do this change. Because now it's unclear if the config apply is failing due to the `auth_dbname=pgbouncer` in the `[databases]` section or the one in the `[pgbouncer]` section. Since the `test_auth_dbname_usage` already tests `auth_dbname=pgbouncer` in the `[databases]` section, it seems good to remove here. 

```suggestion
        * = host={bouncer.host} port={bouncer.port}
```

Reply:
Thank you so much for your clarification! That's right. I have pushed the changes related to that.